[22563] Reload and Merge WP table in background

oliverguenther opened the pull request:
This extracts query / API requests from wp-list controller to a separate service that is instrumented in the background by wp-list whenever an inline-edit operation has completed.

Based on the new results from the API, it will try to identifiy work packages that have not changed (by looking for identical IDs with the same lockVersion) and replace values from the API with existing entries to retain all loaded links / resources.

As we need an option to disable caching for specific routes, or ideally,
specify other means of caching requests, this PR also adds:
- A wrapper for $http to catch caching options passed through
  Restangular
- An improved CacheService that can handle loading promises
  (previously, it interfaced only to Resources)
#### TODOS
- [x] Provide some kind of animation to lost rows

https://community.openproject.org/work_packages/22563/activity

oliverguenther committed:
Separate wp-list controller query loading into services

oliverguenther committed:
Remove PaginationService setting from params

oliverguenther committed:
Very crudely reload table

oliverguenther committed:
Copy live work packages based on lockVersion

oliverguenther committed:
Use restangular#withConfig to initialize uncached provider

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-inline-edit-reload-table.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-inline-edit-reload-table.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Fix schema loading

oliverguenther committed:
Wrap $http for caching

We'll need an option to disable caching for specific routes, or ideally,
specify other means of caching requests.

This commit adds:
  - A wrapper for $http to catch caching options passed through
  Restangular
  - An improved CacheService that can handle loading promises
  (previously, it interfaced only to Resources)

oliverguenther committed:
Very basic animation for removed rows

oliverguenther committed:
Disable cache for queries

oliverguenther committed:
Cache resources in halResource

oliverguenther committed:
Remove uncachedProvider

We'll no longer need that one

oliverguenther committed:
Keep loading Promise in Hal resource

oliverguenther commented:
@furinvader Please take a look at these changes as soon as you're ready.

furinvader commented on frontend/app/components/api/api-v3/hal/hal-resource.service.ts:
```diff
@@ -85,16 +86,22 @@ function halResource(halTransform, HalLink, $q) {
       }
 
-      return $q.when(this);
+      this._loadingPromise = this.$links.self().then(resource => {
```
Did you add the `loadingPromise` attribute because of error handling?
I am surely unaware of something here, but for now I don't see the difference between this code and the old one.

oliverguenther replied:
While the schema is not $loaded, calling `$load` multiple will cause multiple requests / transforms to be made. If you return the first promise instead, it is saving all those requests (and they do happen quite a lot, once for each field)

furinvader committed:
Replace _$loadingPromise with $self and make it public

furinvader commented:
Looks good to me.
Merge, if there is nothing to add.
